File: src/storage/specs.cc
#include "common/dtypes.h"
#include "debug.h"
#include "storage/specs.h"

namespace gear::storage {
ColumnSpec::ColumnSpec(std::vector<size_t> shape, DataType dtype,
                       std::string name)
    : shape(shape), dtype(dtype), name(name) {}

size_t ColumnSpec::size() const {
  size_t elem_size = gear::common::sizeof_dtype(this->dtype);
  return std::accumulate(std::begin(this->shape), std::end(this->shape),
                         elem_size, std::multiplies<size_t>());
}

TableSpec::TableSpec(size_t rank, size_t worldsize, size_t trajectory_length,
                     size_t capacity, size_t num_columns,
                     std::vector<ColumnSpec> column_specs)
    : rank(rank), worldsize(worldsize), trajectory_length(trajectory_length),
      capacity(capacity), num_columns(num_columns), column_specs(column_specs) {
  GEAR_ASSERT(column_specs.size() == num_columns,
              "column specs size mismatch with declared number of columns: "
              "spec num %ld, expected %ld",
              column_specs.size(), num_columns);
  size_t total_size = 0;
  for (const ColumnSpec &cspec : this->column_specs) {
    total_size += cspec.size();
  }
  this->nbytes = total_size * this->trajectory_length * this->capacity;
}

ssize_t TableSpec::index(const std::string &name) {
  ssize_t cid = 0;
  for (const ColumnSpec &cspec : this->column_specs) {
    if (cspec.name == name)
      return cid;
    else
      cid++;
  }
  return -1;
}

size_t TableSpec::size() { return this->nbytes; }

} // namespace gear::storage

File: gear/specs.py
from typing import Any, Dict, Sequence, Union

import libgear.core as glibc
import torch
from gear.dtypes import CVT_DTYPES_TORCH_TO_GEAR, DataType


class ColumnSpec:
    @staticmethod
    def create(
        shape: Sequence[int],
        dtype: torch.dtype = torch.float32,
        name: str = "",
        *args,
        **kwargs
    ):
        if isinstance(dtype, DataType):
            dtype = dtype
        else:
            dtype = CVT_DTYPES_TORCH_TO_GEAR[dtype]
        return glibc.ColumnSpec(shape, dtype, name)


class TableSpec:
    @staticmethod
    def create(
        rank: int = 0,
        worldsize: int = 1,
        trajectory_length: int = 100,
        capacity: int = 32,
        column_specs: Union[Sequence[Dict[str, Any]], glibc.ColumnSpec] = None,
        *args,
        **kwargs
    ):
        num_columns = 0
        cspecs = []
        if column_specs is not None:
            num_columns = len(column_specs)
            for column_spec in column_specs:
                if isinstance(column_spec, glibc.ColumnSpec):
                    cspecs.append(column_spec)
                else:
                    cspecs.append(ColumnSpec.create(**column_spec))
        return glibc.TableSpec(
            rank, worldsize, trajectory_length, capacity, num_columns, cspecs
        )
